### PR TITLE
Set MaxConcurrentStreams = 100 on new Http2Connection until first SETTINGS frame

### LINKED ISSUE #40249: SocketsHttpHandler.EnableMultipleHttp2Connections exceeds server stream limit
### Description

SocketsHttpHandler exceeds the maximum stream limit when EnableMultipleHttp2Connections is true and 250 requests are created simultaneously.

You can see the server reach 100 streams and then receive another new stream here:

```
0.139s SERVER Microsoft.AspNetCore.Server.Kestrel - Trace: Connection id "0HM1N977N3S3G" received HEADERS frame for stream ID 199 with length 164 and flags END_HEADERS
0.139s SERVER Microsoft.AspNetCore.Server.Kestrel - Debug: Connection id "0HM1N977N3S3G" reached the maximum number of concurrent HTTP/2 streams allowed.
0.139s SERVER Microsoft.AspNetCore.Server.Kestrel - Trace: Connection id "0HM1N977N3S3G" received HEADERS frame for stream ID 201 with length 164 and flags END_HEADERS
0.141s SERVER Microsoft.AspNetCore.Server.Kestrel - Debug: Connection id "0HM1N977N3S3G": HTTP/2 stream error.
Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2StreamErrorException: HTTP/2 stream ID 201 error (REFUSED_STREAM): A new stream was refused because this connection has reached its stream limit.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.StartStream()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.DecodeHeadersAsync(Boolean endHeaders, ReadOnlySequence`1& payload)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.ProcessHeadersFrameAsync[TContext](IHttpApplication`1 application, ReadOnlySequence`1& payload)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.ProcessFrameAsync[TContext](IHttpApplication`1 application, ReadOnlySequence`1& payload)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.ProcessRequestsAsync[TContext](IHttpApplication`1 application)
```

Complete log: [http2streamerror.log](https://github.com/dotnet/runtime/files/5013382/http2streamerror.log)

### Configuration

SDK 5.0.100-rc.1.20401.3

### Regression?

Yes. Also happens when EnableMultipleHttp2Connections is the default value (3.1 behavior)

### Other information

Test where I observed this error: https://github.com/JamesNK/grpc-dotnet/commit/9fbef7c8f5455df0990bf4f4f63ce909b055fdd2

@alnikola

### PR DESCRIPTION
**Pull request by alnikola**
Http2Connection is currently created with an 'infinite' MaxConcurrentStreams value as it's demanded by RFC. However, the most of servers does not support unbound active streams number on an HTTP/2 connection because it would make them vulnerable against DOS attacks. Thus, they send a specific MaxConcurrentStreams value with the first SETTINGS frame and respond with REFUSE_STREAM error to opening extra streams attempts. Correct processing this error on client side gets tricky in request burst scenarios due to a very asynchronous nature of HTTP/2. In such cases, client can start sending DATA frames on streams above the limit which make the server to respond with GOAWAY terminating the connection.

Proper fix of the aforementioned issue would be pretty complex, so this PR applies a simple workaround by setting MaxConncurrentStreams to 100 instead of 'infinite' in Http2Connection constructor.

Fixes #40249

### CODE REVIEW

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs
Comment (Tratcher) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:269]:
```suggestion
                    ProcessSettingsFrame(frameHeader, initialFrame: true);
```

**Review by Tratcher — APPROVED** [review]

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:300]: I don't understand why this change is necessary. Same question for the identical change in the below test.

**Review by geoffkizer — CHANGES_REQUESTED** [review]
We should add a test where we hold off on sending an initial SETTINGS frame and validate that the connection creates exactly 100 streams and then queues the 101st. Then send an initial empty SETTINGS frame and confirm that this unblocks the connection to send the 101st.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs
Comment (JamesNK) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:616]: Should the call here be to ChangeMaxConcurrentStreams? That is what happens when `SettingId.MaxConcurrentStreams` setting is received.

**Review by JamesNK — COMMENTED** [review]
I think some extra tests are required:
* There is a burst of requests that exceeds 100 and enable multiple HTTP/2 connections is enabled. I would expect multiple connections to be created, each up to the 100 limit
* There is a burst of requests that exceeds 100 and the server then returns a lower settings value. I would expect pending requests to stay queued until the lower setting value is reached (Kestrel [now better supports](https://github.com/dotnet/aspnetcore/pull/24828) very low max concurrent settings values)
* There is a burst of requests that exceeds 100 and the server then returns a higher settings value. I would expect queued requests would immediately be sent. I think this is the test that @geoffkizer mentioned
* There is a burst of requests that exceeds 100 and the server then returns no settings value, and so the limit is infinite. I would expect queued requests would immediately be sent.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs
Comment (geoffkizer) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:616]: Yes, it should. Thanks for catching that.

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs
Comment (alnikola) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs]: It was a mistake due to my misunderstanding. Fixed.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs
Comment (alnikola) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs]: Fixed.

**Comment by alnikola** [comment]
@geoffkizer I added 3 more tests, including the one you requested above.
Could you please review and approve if everything is fine?
Please, also **merge the PR** after the final approval.

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:371]:
```suggestion
            const int DefaultMaxConcurrentStreams = 100;
```

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:368]: Since these tests are dependent on delays, they should be Outerloop.

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs
Comment (geoffkizer) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:371]: (and elsewhere below)

**Comment by geoffkizer** [comment]
Minor test issues above.

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs
Comment (alnikola) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:369]: Fixed

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs
Comment (alnikola) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs]: All occurrences fixed.

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:424]: Here too :)

Sorry, missed this one before.

Comment (alnikola) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs]: Fixed

**Review by geoffkizer — APPROVED** [review]

**Pull request merged**
